flocker/ca/functional/test_validation.py

```diff
@@ -0,0 +1,274 @@
+            self.result.errback(reason)
+            self.result = None
+
+
+class CredentialSet(PRecord):
+    """
+    A full set of credentials for a CA.
+
+    :param RootCredential root: The CA root credential.
+    :param ControlCredential control: A control service credential.
+    :param UserCredential user: A user credential.
+    :param NodeCredential node: A CA root credentials.
+    """
+    root = field()
+    control = field()
+    user = field()
+    node = field()
+
+    @staticmethod
+    def create():
+        """
+        :return: A new ``CredentialSet``.
+        """
+        directory = FilePath(mkdtemp())
+        atexit.register(rmtree, directory.path)
+        root = RootCredential.initialize(directory, b"mycluster")
+        user = UserCredential.initialize(directory, root, u"allison")
+        node = NodeCredential.initialize(directory, root)
+        control = ControlCredential.initialize(directory, root, b"127.0.0.1")
+        return CredentialSet(root=root, user=user, node=node, control=control)
+
+
+def get_sets():
+    """
+    :return: Tuple, a pair of ``CredentialSet`` instances.
+    """
+    a = CredentialSet.create()
+    b = CredentialSet.create()
+    global get_sets
+
+    def get_sets():
+        return a, b
+    return get_sets()
+
+
+def make_validation_tests(context_factory_fixture,
+                          good_certificate_name,
+                          validator_is_client):
+    """
+    Create a ``TestCase`` for the validator of a specific certificate type.
+
+    :param context_factory_fixture: Create a context factory that
+         implements the required validation given a ``CredentialSet``,
+         given a port number on localhost and a ``CertificateSet``. For
+         server context factories the port number can be ignored.
+
+    :param str good_certificate_name: Name of certificate (an attribute of
+        ``CredentialSet``) that should validate successfully.
+
+    :param bool validator_is_client: Whether or not the context factory
+         being tested is a client.
+
+    :return TestCase: Tests for given validator.
+    """
+    bad_name, another_bad_name = {"user", "node", "control"}.difference(
+        {good_certificate_name})
+
+    class ValidationTests(TestCase):
+        """
+        Tests to ensure correct validation of a specific type of certificate.
+
+        :ivar CertificateSet good_ca: The certificates for the CA we expect.
+
+        :ivar CertificateSet another_ca: A different CA's certificates.
+        """
+        def setUp(self):
+            self.good_ca, self.another_ca = get_sets()
+
+        def assert_validates(self, credential):
+            """
+            Asserts that a TLS handshake is successfully established between a
+            client using the validation logic and a server based on the
+            given credential.
+
+            :param credential: The high-level credential to use.
+
+            :return ``Deferred``: Fires on success.
```

Comment:
Trial catches it and fails? I guess given this code's importance I should be more explicit.